proposals/ssss-distributed-actor-runtime.md

```diff
@@ -0,0 +1,1798 @@
+      //   MARK INITMAP[IMPLICIT_HOP_TO_SELF] = DONE
+      // }
+      //
+      // IF INITMAP[ACTOR_READY] != INITIALIZED {
+      //   << system.actorReady(self)
+      //   MARK INITMAP[ACTOR_READY] = INITIALIZED
+      // }
+      
+      loops -= 1
+    }
+  }
+}
+```
+
+Using this technique we are able to emit the ready call only once, and put off the complexity of dealing with repeated ready calls from distributed actor system library authors. The same technique is used to avoid hopping to the self executor 10 times, but instead the implicit hop-to-self is only performed once, on the initial iteration where the actor became fully initialized.
+
+Things get more complicated in face of failable as well as throwing initializers. Specifically, because we not only have to assign identities, we also need to ensure that they are resigned when the distributed actor is deallocated. In the simple, non-throwing initialization case this is simply done in the distributed actor's `deinit`, however some initialization semantics make this more complicated.
+
+### Resigning Distributed Actor IDs
+
+In addition to assigning `ID` instances to specific actors as they get created, we must also *always* ensure the `ID`s assigned are resigned as their owning actors get destroyed. 
+
+Resigning an `ID` allows the actor system to release any resources it might have held in association with this actor. Most often this means removing it from some internal lookup table that was used to implement the `resolve(ID) -> Self` method of a distributed actor, but it could also imply tearing down connections, clearing caches, or even dropping any in-flight messages addressed to the now terminated actor.
+
+In the simple case this is trivially solved by deinitialization: we completely initialize the actor, and once it deinitializes, we invoke resign the ID in the actor's deinitializer:
+
+```swift
+deinit {
+  // << self.actorSystem.resignID(self.id)
+}
+```
+
+This also works with user defined deinitializers, where the resign call is injected as the *first* operation in the deinitializer:
+
+```swift
+// user-defined deinit
+deinit {
+  // << self.actorSystem.resignID(self.id)
+  print("deinit \(self.id)")
+}
+```
+
+Things get more complicated once we take into account the existence of *failable* and *throwing* initializers. Existing Swift semantics around those types of initializers, and their effect on if and when `deinit` is invoked mean that we need to take special care of them.
+
+Let us first discuss [failable initializers](https://docs.swift.org/swift-book/LanguageGuide/Initialization.html#ID224), i.e. initializers which are allowed to assign `nil` during their initialization. As actors allow such initializers, distributed actors should too, in order to make the friction of moving from local-only to distributed actors as small as possible.
+
+```swift
+distributed actor DA {
+  var int: Int
+  
+  init?(int: Int, system: ActorSystem) [async] {
```

Comment:
Why there are brackets `[]` around `async`?

Reply:
I meant that it's optional, but that probably is confusing -- removing!